# Tugas Hari PERTAMA_PANDAS INTRODUCTION

<h2>Soal 1: Membuat Objek DataFrames</h2>

- Buatlah objek dataframes dari objek dictionary, Bebas!

In [1]:
# Membuat dataframes dari dictionary
import pandas as pd
import numpy as np

dict = {'Negara' : ['Indonesia', 'China', 'Jepang', 'India'], 'Ibu Kota' : ['Jakarta', 'Beijing', 'Tokyo', 'New Delhi'], 'Populasi' : [264, 143, 145, 453]}

df = pd.DataFrame(dict)
print(df)

      Negara   Ibu Kota  Populasi
0  Indonesia    Jakarta       264
1      China    Beijing       143
2     Jepang      Tokyo       145
3      India  New Delhi       453


<h2>Soal 2: Membaca data dari file</h2>

Download file bernama [diamonds.csv](https://drive.google.com/uc?export=download&id=1qQiM9utQUThTPq8jk8stPMh_y2zfqYOd), kemudian baca data tersebut dan assign ke dalam suatu variabel kemudian print.

In [6]:
# Membaca data dan memasukannya kedalam suatu variabel
import pandas as pd
df_from_csv = pd.read_csv('diamonds.csv')


display(df_from_csv)

Unnamed: 0  carat        cut color clarity  depth  table  price     x  \
0               1   0.23      Ideal     E     SI2   61.5   55.0    326  3.95   
1               2   0.21    Premium     E     SI1   59.8   61.0    326  3.89   
2               3   0.23       Good     E     VS1   56.9   65.0    327  4.05   
3               4   0.29    Premium     I     VS2   62.4   58.0    334  4.20   
4               5   0.31       Good     J     SI2   63.3   58.0    335  4.34   
...           ...    ...        ...   ...     ...    ...    ...    ...   ...   
53935       53936   0.72      Ideal     D     SI1   60.8   57.0   2757  5.75   
53936       53937   0.72       Good     D     SI1   63.1   55.0   2757  5.69   
53937       53938   0.70  Very Good     D     SI1   62.8   60.0   2757  5.66   
53938       53939   0.86    Premium     H     SI2   61.0   58.0   2757  6.15   
53939       53940   0.75      Ideal     D     SI2   62.2   55.0   2757  5.83   

          y     z  
0      3.98  2.43  
1      3.84  2.31  
2      4.07  2.31  
3      4.23  2.63  
4      4.35  2.75  
...     ...   ...  
53935  5.76  3.50  
53936  5.75  3.61  
53937  5.68  3.56  
53938  6.12  3.74  
53939  5.87  3.64  

[53940 rows x 11 columns]

Expected Result:

![alt text](https://drive.google.com/uc?id=1BvgM5mJEVfj0fllrjkBfQwn0QIFkry_t)

<h2>Soal 3: DataFrame vs Series</h2>

- Apa perbedaan Series dengan list dan dictionary
- Apa perbedaan dataframe dan series
- Buatlah suatu data series

Jawab: Series adalah suatu objek dengan satu dimensi yang dapat menyimpan jenis data seperti integer, string, dsb. List digunakan pada data yang ingin dibuat struktur data berurutan, jika tidak memerlukan struktur data yang berurutan, maka bisa digunakan dictionary.

Perbedaan data frame dengan series adalah jika series merupakan suatu objek satu dimensi sedangkan data frame merupakan suatu objek dua dimensi dengan menyimpan data lebih terstruktur.

In [7]:
# code here
import pandas as pd

a = pd.Series([4,8,9,7])
print(a)

0    4
1    8
2    9
3    7
dtype: int64


<h2>Soal 4: Pemeriksaan Data Sederhana</h2>

Di soal no 2 kamu telah membaca data tentang diamonds. pada soal ini cobalah suatu metode dari pandas untuk mengetahui beberapa data pertama dan beberapa data terakhir. kemudian sebutkan nilai dari:
- baris pertama dari column price
- baris terakhir dari column color

tunjukan cara kalian menemukan nilai tersebut di block di bawah ini.

In [8]:
# code here

a = df_from_csv['price'].head(1)
b = df_from_csv['color'].tail(1)
display(a)
display(b)

0    326
Name: price, dtype: int64

53939    D
Name: color, dtype: object

<h2>Soal 5: Deskriptif Statistik</h2>

Jelaskan apa itu deskriptis statistik!

Jawab: Deskriptis statistik adalah suatu metode yang berkaitan dengan pengumpulan dan penyajian dari suatu data untuk mendapatkan informasi yang bisa digunakan. Informasi yang didapatkan antara lain seperti ukuran pemusatan data, ukuran penyebaran data. Metode ini hanya digunakan untuk mendeskripsikan kondisi dari suatu data yang dimiliki dan menyajikannya dalam bentuk tabel dan tidak menarik kesimpulan apapun mengenai data tersebut.

<h2>Soal 6: Practice Deskriptif Statistik</h2>

- Gunakan suatu metode dari pandas untuk mengetahui deskriptif statistik dari suatu data
- Berapa nilai rata-rata dari column price
- Berapa nilai standar deviasi dari column depth
- Berapa nilai maximum dari column carat

In [12]:
import statistics

In [19]:
# code here
df = pd.read_csv('diamonds.csv')
df.describe()

Unnamed: 0         carat         depth         table         price  \
count  53940.000000  53940.000000  53940.000000  53940.000000  53940.000000   
mean   26970.500000      0.797940     61.749405     57.457184   3932.799722   
std    15571.281097      0.474011      1.432621      2.234491   3989.439738   
min        1.000000      0.200000     43.000000     43.000000    326.000000   
25%    13485.750000      0.400000     61.000000     56.000000    950.000000   
50%    26970.500000      0.700000     61.800000     57.000000   2401.000000   
75%    40455.250000      1.040000     62.500000     59.000000   5324.250000   
max    53940.000000      5.010000     79.000000     95.000000  18823.000000   

                  x             y             z  
count  53940.000000  53940.000000  53940.000000  
mean       5.731157      5.734526      3.538734  
std        1.121761      1.142135      0.705699  
min        0.000000      0.000000      0.000000  
25%        4.710000      4.720000      2.910000  
50%        5.700000      5.710000      3.530000  
75%        6.540000      6.540000      4.040000  
max       10.740000     58.900000     31.800000

In [20]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 53940 entries, 0 to 53939
Data columns (total 11 columns):
 #   Column      Non-Null Count  Dtype  
---  ------      --------------  -----  
 0   Unnamed: 0  53940 non-null  int64  
 1   carat       53940 non-null  float64
 2   cut         53940 non-null  object 
 3   color       53940 non-null  object 
 4   clarity     53940 non-null  object 
 5   depth       53940 non-null  float64
 6   table       53940 non-null  float64
 7   price       53940 non-null  int64  
 8   x           53940 non-null  float64
 9   y           53940 non-null  float64
 10  z           53940 non-null  float64
dtypes: float64(6), int64(2), object(3)
memory usage: 4.5+ MB


In [25]:
rerata_price = df['price'].mean()
print('Rerata dari kolom price = ', rerata_price)

Rerata dari kolom price =  3932.799721913237


Jawab:
Rerata dari kolom price yaitu 3932.799722

Standar deviasi dari kolom depth yaitu 1.432621	

Nilai maximum dari kolom carat yaitu 5.010000

# Tugas Hari KEDUA_AKSES/INDEXING DAN TRANSFORMASI DATA

In [ ]:
# Persiapan membuat data
import pandas as pd
import random

pelajaran = ['Matematika', 'Fisika', 'Biologi', 'Kimia', 'Geologi', 'Ekonomi']
kelas = ['12A', '12B', '12C', '12D']

hasil_nilai = {
    'siswa_id' : range(1, 21),
    'pelajaran' : [random.choice(pelajaran) for i in range(20)],
    'nilai' : [random.choice(range(0, 100)) for i in range(20)],
    'jenis-kelamin' : [random.choice(['laki-laki', 'perempuan']) for i in range(20)],
    'kelas' : [random.choice(kelas) for i in range(20)]
}

df = pd.DataFrame(hasil_nilai)
df.head()

siswa_id pelajaran  nilai jenis-kelamin kelas
0         1   Ekonomi     30     laki-laki   12A
1         2   Geologi     89     perempuan   12C
2         3   Geologi     96     laki-laki   12A
3         4   Ekonomi     48     perempuan   12D
4         5   Ekonomi     18     laki-laki   12A

<h2>Soal 1: Indexing</h2>

1. Lakukan indexing terhadap dataframe untuk mengakses column nilai dan kelas saja
2. Lakukan indexing terhadap dataframe untuk mengakses row 5 sampai 10 beserta column pelajaran dan nilai

In [ ]:
# code here
#1. Indexing untuk mengakses column nilai dan kelas saja
import pandas as pd
df.loc[:,['nilai', 'kelas']]

nilai kelas
0      30   12A
1      89   12C
2      96   12A
3      48   12D
4      18   12A
5      84   12A
6      35   12D
7      49   12A
8      32   12A
9      78   12B
10     12   12D
11      6   12B
12     41   12C
13     71   12D
14     75   12B
15      9   12A
16     64   12C
17     92   12A
18     45   12C
19     93   12D

In [ ]:
# 2. Indexing untukmengakses row 5 sampai 10 beserta column pelajaran dan nilai
import pandas as pd

df.loc[:,['pelajaran', 'nilai']][4:10]

pelajaran  nilai
4     Ekonomi     18
5       Kimia     84
6     Ekonomi     35
7  Matematika     49
8     Biologi     32
9     Ekonomi     78

<h2>Soal 2: Filter DataFrame Dengan Boolean</h2>

1. Tampilkan semua data siswa pada mata pelajaran Ekonomi dan Fisika
2. Tampilkan semua data siswa yang mempunyai nilai lebih dari 50 di mata pelajaran Kimia 
3. Tampilkan semua data yang mempunyai nilai lebih dari 70, berjnis kelamin perempuan dan di kelas 12C

In [ ]:
# code here
#Data siswa pada pelajaran Ekonomi dan Fisika
pelajaran = df[(df['pelajaran']=='Ekonomi')|(df['pelajaran']=='Fisiska')]
pelajaran

siswa_id pelajaran  nilai jenis-kelamin kelas
0          1   Ekonomi     30     laki-laki   12A
3          4   Ekonomi     48     perempuan   12D
4          5   Ekonomi     18     laki-laki   12A
6          7   Ekonomi     35     laki-laki   12D
9         10   Ekonomi     78     laki-laki   12B
12        13   Ekonomi     41     laki-laki   12C

In [ ]:
# 2 Data siswa yang mempunyai nilai lebih dari 50 di mata pelajaran Kimia 
kimia= df[(df['pelajaran']=='Kimia') & (df['nilai']>50)]
kimia

siswa_id pelajaran  nilai jenis-kelamin kelas
5          6     Kimia     84     laki-laki   12A
13        14     Kimia     71     perempuan   12D
16        17     Kimia     64     laki-laki   12C

In [ ]:
# Data yang mempunyai nilai lebih dari 70, berjenis kelamin perempuan dan di kelas 12C
data= df[(df['nilai']>70) & (df['jenis-kelamin']=='perempuan')& (df['kelas']=='12C')]
data

siswa_id pelajaran  nilai jenis-kelamin kelas
1         2   Geologi     89     perempuan   12C

<h2>Soal 3: Data Transformation</h2>

lakukan data transformasi terhadap column 'nilai' dengan fungsi transformasi, berikut rumus dari normalisasi

![alt text](https://drive.google.com/uc?id=1wed-EbRSTA0JP1ceDp6lcKBqH2yJ4aa3)

In [ ]:
# code here
import pandas as pd

def X_normalized(x):
    return (x - x.min())/(x.max()-x.min())

print(df[['nilai']].apply(X_normalized))

       nilai
0   0.266667
1   0.922222
2   1.000000
3   0.466667
4   0.133333
5   0.866667
6   0.322222
7   0.477778
8   0.288889
9   0.800000
10  0.066667
11  0.000000
12  0.388889
13  0.722222
14  0.766667
15  0.033333
16  0.644444
17  0.955556
18  0.433333
19  0.966667


# Tugas Hari KETIGA_DATA WRANGLING

<h2>Soal 1: Pivoting and Melting</h2>

Jelaskan itu pivoting dan melting data

jawaban: Pivoting adalah suatu pengubahan bentuk data dengan memutar data yang terletak di baris menjadi column. Sedangkan melting adalah suatu pengubahan data dengan memutar dari yang tadinya berada di posisi column menjadi di posisi row.

<h2>Soal 2: Data structure Comprehension</h2>

Data yang rapi adalah:
1. Setiap row menunjukan suatu unit observasi
2. Setiap column merupakan suatu unit variable

Bagaimana pun, dalam menentukan unit observasi itu tergantung dari kasus yang di hadapi. sebagai seorang data scientist pasti akan menemukan berbagai bentuk struktur data. adalah sangat penting bagi seorang data scientist untuk dapat memahami dan menjelaskan setiap struktur data. Kemampuan memahami hubungan antara index baris, column dan isi data adalah kunci untuk memahami informasi apa yang data sedang beritahukan kepada kita.

Di bawah ini terdapat 2 data yang berisi sama, tetapi memiliki struktur yang berbeda. Jelaskan perbedaan dari 2 struktur data tersebut

![alt text](https://drive.google.com/uc?id=1h13CDPzIjc-2XyRcyu-p1fet0Ep1gdMA)

![alt text](https://drive.google.com/uc?id=12ji98ell5tNcCxeBzl4v8qqBtFMgDSr0)

Jawab : Data pertama terlihat lebih rapi dibandingkan dengan data kedua, karena memberikan informasi yang jelas. Dapat dilihat bahwa indeks barisnya berupa Tipe Ujian, setiap kolomnya mewakili mata pelajaran, dan nilai yang tertulis di dalam tabel berisi rata-rata waktu belajar (dalam menit), penyajiannya lebih efisien. Sedangkan pada Data kedua, menurut saya data yang disajikan tidak langsung to the point.

<h2>Soal 3: Reshaping Data 1</h2>

In [ ]:
import pandas as pd
import random

data = {
    'Tipe Ujian' : ['UTS1', 'UAS1', 'UTS2', 'UAS2']*3,
    'Mata Pelajaran' : ['Fisika']*4 + ['Matematika']*4+['Statistika']*4,
    'Rata2 Waktu Belajar (menit)' : [random.choice(range(200, 800)) for i in range(12)],
    'Rata2 Nilai Ujian' : [random.choice(range(40, 100)) for i in range(12)] 
}
df = pd.DataFrame(data)
df

Tipe Ujian Mata Pelajaran  Rata2 Waktu Belajar (menit)  Rata2 Nilai Ujian
0        UTS1         Fisika                          274                 86
1        UAS1         Fisika                          667                 98
2        UTS2         Fisika                          487                 71
3        UAS2         Fisika                          564                 88
4        UTS1     Matematika                          664                 80
5        UAS1     Matematika                          750                 91
6        UTS2     Matematika                          556                 40
7        UAS2     Matematika                          409                 49
8        UTS1     Statistika                          588                 84
9        UAS1     Statistika                          696                 76
10       UTS2     Statistika                          726                 62
11       UAS2     Statistika                          575                 62

Perhatikan data di atas, kemudian ubahlah bentuknya agar sesuai yang di harapkan.

In [ ]:
# code here

pivot_df= pd.pivot(df, index= 'Tipe Ujian', columns= 'Mata Pelajaran', values= ['Rata2 Nilai Ujian','Rata2 Waktu Belajar (menit)'])

pivot_df

Rata2 Nilai Ujian                        \
Mata Pelajaran            Fisika Matematika Statistika   
Tipe Ujian                                               
UAS1                          98         91         76   
UAS2                          88         49         62   
UTS1                          86         80         84   
UTS2                          71         40         62   

               Rata2 Waktu Belajar (menit)                        
Mata Pelajaran                      Fisika Matematika Statistika  
Tipe Ujian                                                        
UAS1                                   667        750        696  
UAS2                                   564        409        575  
UTS1                                   274        664        588  
UTS2                                   487        556        726

Expected Result:

![alt text](https://drive.google.com/uc?id=1XufsfqZe7-XZLsb92SpATexy-QRZE0ow)

<h2>Soal 4: Reshaping Data 2</h2>

In [ ]:
import pandas as pd

data = {
    'Hari' : ['Sabtu', 'Minggu'],
    'Kebun Binatang' : [271, 399],
    'Dufan' : [501, 700],
    'Disney' : [1000, 1001],
    'Bali' : [900, 803]
}
df = pd.DataFrame(data).set_index('Hari')
df

Kebun Binatang  Dufan  Disney  Bali
Hari                                       
Sabtu              271    501    1000   900
Minggu             399    700    1001   803

diatas adalah data tentang jumlah pengunjung di berbagai tempat hiburan, ubahlah struktur data tersebut sehingga sesuai dengan hasil yang diharapkan.

In [ ]:
# Mengubah struktur data sehingga sesuai degan hasil yang diharapkan
pd.melt(df.reset_index(),id_vars =['Hari'], value_vars =['Kebun Binatang', 'Dufan','Disney','Bali'], 
        var_name='Tempat Hiburan',value_name='Pengunjung')

Hari  Tempat Hiburan  Pengunjung
0   Sabtu  Kebun Binatang         271
1  Minggu  Kebun Binatang         399
2   Sabtu           Dufan         501
3  Minggu           Dufan         700
4   Sabtu          Disney        1000
5  Minggu          Disney        1001
6   Sabtu            Bali         900
7  Minggu            Bali         803

Expected Result:

![alt text](https://drive.google.com/uc?id=159mIqcyfeD30EEcyZmmkovORa6cZDLtn)

<h2>Soal 5: Pivoting and Aggregating</h2>

In [ ]:
import pandas as pd

data = {
    'Tipe Ujian' : ['UTS1', 'UAS1']*3,
    'Mata Pelajaran' :['Statistika', 'Fisika', 'Fisika', 'Statistika', 'Fisika', 'Statistika'],
    'Jumlah Peserta' : [14, 17, 12, 16, 14, 13]
}
df = pd.DataFrame(data).set_index('Tipe Ujian')
df


Mata Pelajaran  Jumlah Peserta
Tipe Ujian                               
UTS1           Statistika              14
UAS1               Fisika              17
UTS1               Fisika              12
UAS1           Statistika              16
UTS1               Fisika              14
UAS1           Statistika              13

In [ ]:
#Mengubah data sesuai yang diharapkan
pivot_table_df= pd.pivot_table(df, index='Tipe Ujian', columns='Mata Pelajaran', values='Jumlah Peserta', aggfunc='sum')

pivot_table_df

Mata Pelajaran  Fisika  Statistika
Tipe Ujian                        
UAS1                17          29
UTS1                26          14

Perhatikan bahwa data jumlah peserta ujian untuk suatu tipe ujian dan mata pelajaran tertentu ada yang duplikat, ubahlah struktur data tersebut dan lakukan aggregasi berjenis jumlah sehingga sesuai dengan hasil yang di harapkan

Expected Result:

![alt text](https://drive.google.com/uc?id=1emoz7orYR-0mCHffvAgM8j-V1P6yVHf0)

# Tugas Hari KEEMPAT_MARGING DAN GROUPBY

<b>Note</b>: Pada kuis kali ini kita akan menggunakan data diamonds.csv yang bisa kalian download [disini](https://drive.google.com/uc?export=download&id=1qQiM9utQUThTPq8jk8stPMh_y2zfqYOd). setelah di download, kalian pindahkan data tersebut satu level dengan direktori file notebook ini.


In [ ]:
import pandas as pd
df = pd.read_csv('diamonds.csv')
df.drop('Unnamed: 0', axis=1, inplace=True)

In [ ]:
df.head()

carat      cut color clarity  depth  table  price     x     y     z
0   0.23    Ideal     E     SI2   61.5   55.0    326  3.95  3.98  2.43
1   0.21  Premium     E     SI1   59.8   61.0    326  3.89  3.84  2.31
2   0.23     Good     E     VS1   56.9   65.0    327  4.05  4.07  2.31
3   0.29  Premium     I     VS2   62.4   58.0    334  4.20  4.23  2.63
4   0.31     Good     J     SI2   63.3   58.0    335  4.34  4.35  2.75

Columns Metadata:

carat: berat diamond

cut: kualitas diamond

color: warna diamond

clarity: level kemurnian diamonds

depth: tinggi dari diamond tabel

table: lebar  dari diamond tabel

price: harga diamond ($US)

x: panjang

y: lebar

z: kedalaaman

<h2>Soal 1: Process Grouping</h2>

Jelaskan prosess yang terjadi dalam grouping data menggungkan method <b>groupby()</b> dari pandas.

jawab: 
Akan terdapat beberapa proses yang terjadi secara berurutan, yaitu :
splitting (memisahkan kedalam suatu grup, berdasarkan kriteria tertentu), applying (melakukan suatu operasi terhadap sekumpulan data di grup-grup tersebut), dan combining (menggabungkan data menjadi suatu struktur baru).

<h2>Soal 2: Grouping All Column</h2>
    
Group data diamond berdasarkan cut (kualitas), lakukan aggregasi dengan fungsi rata2.

Kemudian jawab pertanyaan berikut.
- Berapa harga rata2 dari diamond berkualitas Premium

In [ ]:
df.groupby('cut').mean()

carat      depth      table        price         x         y  \
cut                                                                          
Fair       1.046137  64.041677  59.053789  4358.757764  6.246894  6.182652   
Good       0.849185  62.365879  58.694639  3928.864452  5.838785  5.850744   
Ideal      0.702837  61.709401  55.951668  3457.541970  5.507451  5.520080   
Premium    0.891955  61.264673  58.746095  4584.257704  5.973887  5.944879   
Very Good  0.806381  61.818275  57.956150  3981.759891  5.740696  5.770026   

                  z  
cut                  
Fair       3.982770  
Good       3.639507  
Ideal      3.401448  
Premium    3.647124  
Very Good  3.559801

Expected Output:

![alt text](https://drive.google.com/uc?id=1GKanuLzNv_pYBFVRBmQywxQw39uQBvSl)

jawab: 4584.257704

<h2>Soal 3: Grouping Some Column base on Multiple Criteria</h2>
    
Lakukan groupby pada data berdasarkan kriteria cut (kualitas) dan color (warna) secara berurutan, dengan aggregasi pada column carat, table, dan price saja. dengan fungsi agregasi rata-rata.

Kemudian jawab pertanyaan berikut:
- Berapa harga diamond yang berkriteria premium dan berwarna J
- Berapa nilai dari carat diamond yang berkriteria very good dan berwarna H

Expected Output:

Note: gambar hanya sebagaian dari keseluruhan data

![alt text](https://drive.google.com/uc?id=1XB_uHor7ZCPQUH9uQjSfd6aC6_yJnjh9)

In [ ]:
# code here
df.groupby(['cut','color'])[['carat', 'table', 'price']].agg('mean')

carat      table        price
cut       color                                  
Fair      D      0.920123  58.969325  4291.061350
          E      0.856607  59.364732  3682.312500
          F      0.904712  59.453205  3827.003205
          G      1.023822  58.773248  4239.254777
          H      1.219175  58.696370  5135.683168
          I      1.198057  59.237143  4685.445714
          J      1.341176  58.917647  4975.655462
Good      D      0.744517  58.541541  3405.382175
          E      0.745134  58.779957  3423.644159
          F      0.775930  58.910891  3495.750275
          G      0.850896  58.471986  4123.482204
          H      0.914729  58.611111  4276.254986
          I      1.057222  58.773946  5078.532567
          J      1.099544  58.813029  4574.172638
Ideal     D      0.565766  55.965632  2629.094566
          E      0.578401  55.967461  2597.550090
          F      0.655829  55.924203  3374.939362
          G      0.700715  55.902375  3720.706388
          H      0.799525  55.965843  3889.334831
          I      0.913029  56.021357  4451.970377
          J      1.063594  56.012612  4918.186384
Premium   D      0.721547  58.718964  3631.292576
          E      0.717745  58.779461  3538.914420
          F      0.827036  58.679279  4324.890176
          G      0.841488  58.702360  4500.742134
          H      1.016449  58.792034  5216.706780
          I      1.144937  58.771849  5946.180672
          J      1.293094  58.874752  6294.591584
Very Good D      0.696424  58.041309  3470.467284
          E      0.676317  58.038875  3214.652083
          F      0.740961  57.848429  3778.820240
          G      0.766799  57.784428  3872.753806
          H      0.915948  57.903015  4535.390351
          I      1.046952  58.105150  5255.879568
          J      1.133215  58.277729  5103.513274

jawab:
Harga diamond yang berkriteria premium dan berwarna J yaitu 6294.591584
Nilai dari carat diamond yang berkriteria very good dan berwarna H yaitu 0.915948	

<h2>Soal 4: Group By Multiple Aggregate Base on Multiple Criteria</h2>
    
Lakukan groupby pada data berdasarkan kriteria cut(kualitas) dan color (warna) secara berurutan kemudian aggregasi objek group by nya dengan fungsi rata2 dan median.

Kemudian Jawab Pertanyaan berikut:

- Berapa nilai median dari harga diamond berkualitas good dan berwarna F

Expected Output:

Note : gambar hanya sebagaian dari keseluruhan data

![alt text](https://drive.google.com/uc?id=1fUPIse5ELrzj26ZyWBqsAVOUb6yf7UoD)

In [ ]:
# code here
df.groupby(['cut','color']).agg(['mean','median'])

carat             depth             table         \
                     mean median       mean median       mean median   
cut       color                                                        
Fair      D      0.920123  0.900  64.048466   64.8  58.969325   58.0   
          E      0.856607  0.900  63.319643   64.8  59.364732   59.0   
          F      0.904712  0.900  63.508013   64.8  59.453205   59.0   
          G      1.023822  0.980  64.339809   65.0  58.773248   58.0   
          H      1.219175  1.010  64.585149   65.1  58.696370   58.0   
          I      1.198057  1.010  64.220571   65.1  59.237143   58.0   
          J      1.341176  1.030  64.357143   65.0  58.917647   58.0   
Good      D      0.744517  0.700  62.365710   63.4  58.541541   58.0   
          E      0.745134  0.700  62.203751   63.3  58.779957   58.0   
          F      0.775930  0.710  62.202310   63.2  58.910891   59.0   
          G      0.850896  0.900  62.526636   63.4  58.471986   58.0   
          H      0.914729  0.900  62.499573   63.4  58.611111   58.0   
          I      1.057222  1.000  62.475096   63.5  58.773946   58.0   
          J      1.099544  1.020  62.395765   63.4  58.813029   58.0   
Ideal     D      0.565766  0.500  61.678370   61.8  55.965632   56.0   
          E      0.578401  0.500  61.686677   61.8  55.967461   56.0   
          F      0.655829  0.530  61.675562   61.8  55.924203   56.0   
          G      0.700715  0.540  61.700164   61.8  55.902375   56.0   
          H      0.799525  0.700  61.733194   61.8  55.965843   56.0   
          I      0.913029  0.740  61.793598   61.9  56.021357   56.0   
          J      1.063594  1.030  61.821987   61.9  56.012612   56.0   
Premium   D      0.721547  0.580  61.169245   61.3  58.718964   59.0   
          E      0.717745  0.580  61.175909   61.3  58.779461   59.0   
          F      0.827036  0.760  61.259974   61.5  58.679279   59.0   
          G      0.841488  0.755  61.279480   61.4  58.702360   59.0   
          H      1.016449  1.010  61.322034   61.5  58.792034   59.0   
          I      1.144937  1.140  61.328641   61.6  58.771849   59.0   
          J      1.293094  1.250  61.390099   61.6  58.874752   59.0   
Very Good D      0.696424  0.610  61.750165   62.2  58.041309   58.0   
          E      0.676317  0.570  61.730250   62.0  58.038875   58.0   
          F      0.740961  0.700  61.721627   62.0  57.848429   58.0   
          G      0.766799  0.700  61.841279   62.1  57.784428   58.0   
          H      0.915948  0.900  61.968366   62.3  57.903015   58.0   
          I      1.046952  1.005  61.934551   62.2  58.105150   58.0   
          J      1.133215  1.060  61.902065   62.2  58.277729   58.0   

                       price                 x                y         \
                        mean  median      mean median      mean median   
cut       color                                                          
Fair      D      4291.061350  3730.0  6.018344  6.080  5.963190  6.010   
          E      3682.312500  2956.0  5.909063  6.070  5.858214  6.020   
          F      3827.003205  3035.0  5.990513  6.060  5.931122  5.990   
          G      4239.254777  3057.0  6.173822  6.125  6.114076  6.040   
          H      5135.683168  3816.0  6.579373  6.330  6.497393  6.250   
          I      4685.445714  3246.0  6.564457  6.340  6.493486  6.270   
          J      4975.655462  3302.0  6.747311  6.490  6.675882  6.460   
Good      D      3405.382175  2728.5  5.620076  5.690  5.633897  5.730   
          E      3423.644159  2420.0  5.617889  5.680  5.632454  5.710   
          F      3495.750275  2647.0  5.693443  5.760  5.709659  5.790   
          G      4123.482204  3340.0  5.850264  6.050  5.862824  6.070   
          H      4276.254986  3468.5  5.968789  6.110  5.974245  6.130   
          I      5078.532567  3639.5  6.253544  6.275  6.260536  6.260   
          J      4574.172638  3733.0  6.377003  6.440  6.386026  6.460   
Ideal     D      2629.094

jawab:
nilai median dari harga diamond berkualitas good dan berwarna F yaitu 63.2

<h2>Soal 5: Group By With Different Aggregation for specific Column</h2>
    
Lakukan groupby pada data berdasarkan criteria cut(kualitas), kemudian aggreagate dengan fungsi max pada column price dan fungsi min pada column carat.

Kemudian Jawab Pertanyaan berikut:

- Berapa nilai max dari price untuk diamond berkualitas Ideal

Expected Output:

![alt text](https://drive.google.com/uc?id=1B_z5PBkLe4hNJJJV3JgDGP9G--llytKf)

In [ ]:
# code here
df.groupby('cut').agg({'price' : 'max', 'carat':'min'})

price  carat
cut                    
Fair       18574   0.22
Good       18788   0.23
Ideal      18806   0.20
Premium    18823   0.20
Very Good  18818   0.20

jawab:
Nilai max dari price untuk diamond berkualitas Ideal yaitu 18806

<h2>Soal 6: Group By With Custom Function</h2>
    
Lakukan groupby pada data berdasarkan criteria cut(kualitas), kemudian lakukan aggregasi dengan fungsi standar deviasi pada column carat dan fungsi rentang nilai antara max dan min pada column price.

Kemudian jawab pertanyaan berikut:
- Berapa selisih harga tertinggi dan terendah dari diamond berkualitas Premium

Expected Output:

![alt text](https://drive.google.com/uc?id=1bl6_8KryXCxLfzpAf4YF80Qvt15a3uQ6)

In [ ]:
# code here
def selisih(x):
    return x.max() - x.min()
df.groupby('cut').agg({'price':selisih,'carat':'std'})

price     carat
cut                       
Fair       18237  0.516404
Good       18461  0.454054
Ideal      18480  0.432876
Premium    18497  0.515262
Very Good  18482  0.459435

jawab:
selisih harga tertinggi dan terendah dari diamond berkualitas Premium yaitu 18497

# Tugas Hari KELIMA_TIME SERIES

<h2>Soal 1: Penggabungan Data</h2>

- Sebutkan perbedaan antara full outer join, inner join, dan left join

Jawab:
Full outer join, digunakan untuk menampilkan gabungan dari semua data. Inner join, digunakan untuk menampilkan semua data yang sama. Sedangkan left join, digunakan untuk menampilkan semua data yang dikirim ke tabel yang tidak berhubungan dan data yang kosong akan bernilai "null"


<h2>Soal 2: Penggabungan Data</h2>

Pada quiz no 2 ini gunakan data [names1881.csv](https://drive.google.com/uc?export=download&id=1Pwu3cv93hI79jmi52kJ05xhophknDKat) dan [names1981.csv](https://drive.google.com/uc?export=download&id=1pvsGaybr423hJXcN35e5nS3sg4ha4jo3), data ini berisi nama bayi yang lahir pada tahun 1881 atau 1981 di Amerika.

- Baca data tersebut dengan memberikan nama column 'Name', 'Gender', 'Count'.
- Tunjukan jumlah dimensi dari data 'names1881.csv' dan 'names1981.csv'
- Gabungkan kedua data tersebut berdasarkan sumbu baris
- Tunjukan jumlah dimensi dari data setelah di gabungkan.

In [ ]:
import pandas as pd
# read data
df1 = pd.read_csv('names1881.csv',names=['Name', 'Gender', 'Count'])
df2 = pd.read_csv('names1981.csv',names=['Name', 'Gender', 'Count'])

# periksa dimensi dari kedua data
print("Jumlah Dimensi data 'names1881.csv' :")
print(df1.shape)
print("Jumlah Baris data 'names1981.csv' :")
print(df2.shape)

# gabungkan data
df = pd.concat([df1, df2], axis = 0)
print(df)

# periksa dimensi data hasil gabungan
print("Jumlah dimensi data setelah penggabungan : ")
print (df.shape)

Jumlah Dimensi data 'names1881.csv' :
(1935, 3)
Jumlah Baris data 'names1981.csv' :
(19455, 3)
            Name Gender  Count
0           Mary      F   6919
1           Anna      F   2698
2           Emma      F   2034
3      Elizabeth      F   1852
4       Margaret      F   1658
...          ...    ...    ...
19450   Zeferino      M      5
19451    Zerrick      M      5
19452   Zimbabwe      M      5
19453     Zoltan      M      5
19454     Zuriel      M      5

[21390 rows x 3 columns]
Jumlah dimensi data setelah penggabungan : 
(21390, 3)


Expected Result: 

Jumlah Dimensi data 'names1881.csv': 

(1935, 3)

Jumlah Baris data 'names1981.csv':

(19455, 3)

![alt text](https://drive.google.com/uc?id=1jMCAdLrcJubjufydZBnzdQkYqDS3Klat)

Jumlah dimensi data setelah penggabungan : 

(21390, 3)

<h2>Soal 3: Penggabungan Data</h2>

Lakukan jenis penggabungan data full outer join, inner join, left outer join dan right outer join pada kedua data berikut: 

In [ ]:
import pandas as pd

data1 = {
    'id' : [1, 2, 3, 4],
    'Feature1' : ['a1', 'a2', 'a3', 'a4'],
    'Feature2' : ['b1', 'b2', 'b3', 'b4']
}

data2 = {
    'id' : [3, 4, 5 ,6, 7],
    'Feature3' : ['c1', 'c2', 'c3', 'c4', 'c5']
}

df1= pd.DataFrame(data1)
df2= pd.DataFrame(data2)

print(df1)
print(df2)

   id Feature1 Feature2
0   1       a1       b1
1   2       a2       b2
2   3       a3       b3
3   4       a4       b4
   id Feature3
0   3       c1
1   4       c2
2   5       c3
3   6       c4
4   7       c5


In [ ]:
# join data
df_outer_join = pd.merge(df1, df2, on='id', how='outer')
print(df_outer_join)

df_inner_join = pd.merge(df1, df2, on='id', how='inner')
print(df_inner_join)

df_left_join = pd.merge(df1, df2, on='id', how='left')
print(df_left_join)

df_right_join = pd.merge(df1, df2, on='id', how='right')
print(df_right_join)

   id Feature1 Feature2 Feature3
0   1       a1       b1      NaN
1   2       a2       b2      NaN
2   3       a3       b3       c1
3   4       a4       b4       c2
4   5      NaN      NaN       c3
5   6      NaN      NaN       c4
6   7      NaN      NaN       c5
   id Feature1 Feature2 Feature3
0   3       a3       b3       c1
1   4       a4       b4       c2
   id Feature1 Feature2 Feature3
0   1       a1       b1      NaN
1   2       a2       b2      NaN
2   3       a3       b3       c1
3   4       a4       b4       c2
   id Feature1 Feature2 Feature3
0   3       a3       b3       c1
1   4       a4       b4       c2
2   5      NaN      NaN       c3
3   6      NaN      NaN       c4
4   7      NaN      NaN       c5


Expected Result:

![alt text](https://drive.google.com/uc?id=1kUpkzWh3Y-X_1H1QXHiccdHf4YjqXsDu)

![alt text](https://drive.google.com/uc?id=1ojoAC2oX9P3ZHl0lS3RXipd7gShLOmmC)

![alt text](https://drive.google.com/uc?id=12jDnQzIosn6dizVwl2nPMX1IhhyEZ4FA)

![alt text](https://drive.google.com/uc?id=11znqscNsVZ-cHY-PPgoHEBDWyc6XRtAL)

<h2>Soal 4: Penggabungan Data berbeda label column Key</h2>

Lakukan jenis penggabungan inner join untuk 2 data yang memiliki nama column key yang berbeda ini:

In [ ]:
data1 = {
    'key_data1' : [1, 2, 3, 4],
    'Feature1' : ['a1', 'a2', 'a3', 'a4'],
    'Feature2' : ['b1', 'b2', 'b3', 'b4']
}

data2 = {
    'key_data2' : [3, 4, 5 ,6, 7],
    'Feature3' : ['c1', 'c2', 'c3', 'c4', 'c5']
}

df1= pd.DataFrame(data1)
df2= pd.DataFrame(data2)

print(df1)
print(df2)

   key_data1 Feature1 Feature2
0          1       a1       b1
1          2       a2       b2
2          3       a3       b3
3          4       a4       b4
   key_data2 Feature3
0          3       c1
1          4       c2
2          5       c3
3          6       c4
4          7       c5


In [ ]:
# Gabungkan kedua data tersebut
df_gabung = pd.merge(df1, df2, left_on='key_data1', right_on='key_data2')
df_gabung

key_data1 Feature1 Feature2  key_data2 Feature3
0          3       a3       b3          3       c1
1          4       a4       b4          4       c2

Expected Result:

![alt text](https://drive.google.com/uc?id=1uBiVOPnqns4al88FDc0ENIeTzIxP4wxH)